### ISSUE TITLE
allow for curve method parameters

### ISSUE THREAD

**Issue by Ken-B** [felipenoris/InterestRates.jl#1]
First of all, thanks for sharing this nice package!

I'm a bit confused by the parametric and interpolation methods in a single final type. The `parameter_id` and `parameter_values` fields can have different meanings dependent on the curve method, is that correct? From my little experience many years ago, I remember some interpolation methods required extra parameters, such as the minimum length or minimum tension splines. I don't see how that fits in the current IRCurve type, maybe a parameters dictionary would be more general?

Anyway, I remember a nice curve method that constructs a yield curve by taking a linear combination of functions on several levels. So on a first level you would make a combination of functions that pass through, say, the 2Y, 5Y and 10Y (the most liquid ones). A second level would then add the combination of function to also match points on, say, 1Y, 3Y, 7Y and 30Y and a third level any remaining liquid points. 
The choice of level maturities could be based on experience or a PCA on rate changes for each maturity. The functions could be a normal, decaying exponential, sigmoid, logistic, constant, sqrt, etc. The choice of functions per level and calibration of the function parameters should be optimized once based on historical/expert sensitivities between different maturities. The linear combination however needs to be solved for every curve.

If you're interested, I could see to add this curve method. I would not some historical data for testing though.

**Comment by felipenoris** [comment]
Hello! Thanks for chiming in!
You´re right. `parameter_id` can have different meanings: for interpolations it holds days to maturity as integers. For term structure models, it holds an identifier for each parameter.
For term structure models, they become useless in terms of the package, but are useful for database storage. In the case of interpolation, we can take advantage of the fact that IRCurve is created with sorted parameters, which leads to a little bit faster interpolations.

I see two possibilities from here: **(1)** we could change function signatures to depend only on AbstractIRCurve, and add _get_ methods for curve attributes. In this case, `IRCurve` would be one possible concrete type to store curve data. **(2)** we could change `IRCurve` data structure to hold a `Dict`, as you suggested. But I would like some time to think about it.

Linear combination of curve methods sounds interesting. Maybe we could create a `CompositeCurveMethod`. Feel free to send a PR on your idea.

Cheers!

**Comment by Ken-B** [comment]
Hi Fepile,

Of course, take some time to think about it. 

I'll see if I can find some historic swap data and try the linear combination of functions method in a PR. 

Cheers,
Ken

**Comment by EricForgy** [comment]
I haven't tried it myself, but for historical swap data, you might find the following useful:
- https://github.com/imanuelcostigan/dataonderivatives

**Comment by felipenoris** [comment]
Testing for performance:
https://github.com/felipenoris/InterestRates.jl/commit/0623b397f996885e7dff22820c26dccab0a35a84
Not sure what's the best design.

**Comment by felipenoris** [comment]
Running **dict-performance.jl** yields:

``` julia
julia> include("dict-performance.jl")
static
  0.297848 seconds (4.40 M allocations: 134.287 MB, 1.92% gc time)
dict any
  1.555316 seconds (4.40 M allocations: 134.277 MB, 0.27% gc time)
dict sym
  0.569182 seconds (4.40 M allocations: 134.277 MB, 1.02% gc time)
```

**Comment by felipenoris** [comment]
As a result from **dict-performance.jl** analysis, changing all curve parameters to a general _Dict_ kills performance.
For now, I decided do go through both solutions I pointed out earlier: (1) functions now depend on _AbstractIRCurve_ allowing for other concrete types, (2) added a _dict_ to _IRCurve_, which is currently unused, to allow for future developments.

So commit https://github.com/felipenoris/InterestRates.jl/commit/38afcf9e48478f5427f7b057f2e3cd248396ca42 closes this issue.

As for historical data, I'm working on an example database, see https://github.com/felipenoris/InterestRates.jl/issues/4. I'll send it upstream when it's ready.

Cheers!
